File: lib/fulink/snapshotHub.ts
/**
 * Fulink 스타일 Evidence 집계 (assets/lib data/snapshot SnapshotHub 로직 포팅)
 * - Evidence 리스트를 bias, longPct, shortPct, consensus, confidence, state로 집계
 */

export type EvidenceKind = 'trend' | 'momentum' | 'flow' | 'pattern' | 'volatility' | 'risk';

export type Evidence = {
  id: string;
  kind: EvidenceKind;
  tf: string;
  score: number;   // -1..+1
  weight: number;  // 0..1
  confidence: number; // 0..1
  meta?: Record<string, unknown>;
};

export type TradeState = 'allow' | 'caution' | 'block';

export type EngineSnapshot = {
  tsMs: number;
  bias: number;
  longPct: number;
  shortPct: number;
  consensus: number;
  confidence: number;
  state: TradeState;
  top: Evidence[];
};

function emptySnapshot(): EngineSnapshot {
  return {
    tsMs: Date.now(),
    bias: 0,
    longPct: 0.5,
    shortPct: 0.5,
    consensus: 0.5,
    confidence: 0.5,
    state: 'caution',
    top: [],
  };
}

/**
 * Evidence 배열을 한 번에 집계해 EngineSnapshot 생성 (Dart SnapshotHub._emit 로직)
 */
export function aggregateToSnapshot(evidences: Evidence[], last?: EngineSnapshot | null): EngineSnapshot {
  const now = Date.now();
  if (!evidences.length) {
    return last ? { ...last, tsMs: now } : emptySnapshot();
  }

  let wSum = 0, sSum = 0, cSum = 0;
  for (const e of evidences) {
    const w = Math.max(0, Math.min(1.2, e.weight * (0.4 + 0.6 * e.confidence)));
    wSum += w;
    sSum += w * Math.max(-1, Math.min(1, e.score));
    cSum += w * Math.max(0, Math.min(1, e.confidence));
  }

  const bias = wSum <= 0 ? 0 : Math.max(-1, Math.min(1, sSum / wSum));
  const conf = wSum <= 0 ? 0.5 : Math.max(0, Math.min(1, cSum / wSum));

  let variance = 0;
  for (const e of evidences) {
    const d = e.score - bias;
    variance += d * d;
  }
  variance /= evidences.length;
  const consensus = Math.max(0, Math.min(1, 1 - variance));

  const state: TradeState =
    conf >= 0.72 && consensus >= 0.62 ? 'allow' :
    conf >= 0.52 && consensus >= 0.45 ? 'caution' : 'block';

  const top = [...evidences]
    .sort((a, b) => {
      const aa = (a.weight * a.confidence) * Math.abs(a.score);
      const bb = (b.weight * b.confidence) * Math.abs(b.score);
      return bb - aa;
    })
    .slice(0, 8);

  return {
    tsMs: now,
    bias,
    longPct: Math.max(0, Math.min(1, (bias + 1) / 2)),
    shortPct: 1 - Math.max(0, Math.min(1, (bias + 1) / 2)),
    consensus,
    confidence: conf,
    state,
    top,
  };
}

File: backups/server-pull-20260822-132009/extract/ailongshort/lib/data/snapshot/snapshot_hub.dart
import 'dart:async';
import 'evidence.dart';
import 'engine_snapshot.dart';

class SnapshotHub {
  final Duration tick;
  final _ev = <Evidence>[];
  final _out = StreamController<EngineSnapshot>.broadcast();
  Timer? _timer;

  EngineSnapshot _last = EngineSnapshot.empty();
  EngineSnapshot get last => _last;
  Stream<EngineSnapshot> get stream => _out.stream;

  SnapshotHub({this.tick = const Duration(seconds: 1)});

  void start() => _timer ??= Timer.periodic(tick, (_) => _emit());

  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  void dispose() {
    stop();
    _out.close();
  }

  void push(Evidence e) => _ev.add(e);

  void _emit() {
    if (_out.isClosed) return;
    final now = DateTime.now().millisecondsSinceEpoch;
    final ev = List<Evidence>.from(_ev);
    _ev.clear();

    if (ev.isEmpty) {
      _last = EngineSnapshot(
        tsMs: now,
        bias: _last.bias,
        longPct: _last.longPct,
        shortPct: _last.shortPct,
        consensus: _last.consensus,
        confidence: _last.confidence,
        state: _last.state,
        top: _last.top,
      );
      _out.add(_last);
      return;
    }

    double wSum = 0, sSum = 0, cSum = 0;
    for (final e in ev) {
      final w = (e.weight * (0.4 + 0.6 * e.confidence)).clamp(0.0, 1.2);
      wSum += w;
      sSum += w * e.score.clamp(-1.0, 1.0);
      cSum += w * e.confidence.clamp(0.0, 1.0);
    }

    final bias = (wSum <= 0 ? 0.0 : (sSum / wSum)).clamp(-1.0, 1.0);
    final conf = (wSum <= 0 ? 0.5 : (cSum / wSum)).clamp(0.0, 1.0);

    double variance = 0;
    for (final e in ev) {
      final d = (e.score - bias);
      variance += d * d;
    }
    variance = variance / ev.length;
    final consensus = (1.0 - variance).clamp(0.0, 1.0);

    final longPct = ((bias + 1) / 2).clamp(0.0, 1.0);
    final shortPct = 1.0 - longPct;

    final state =
        (conf >= 0.72 && consensus >= 0.62) ? TradeState.allow :
        (conf >= 0.52 && consensus >= 0.45) ? TradeState.caution :
        TradeState.block;

    ev.sort((a, b) {
      final aa = (a.weight * a.confidence) * a.score.abs();
      final bb = (b.weight * b.confidence) * b.score.abs();
      return bb.compareTo(aa);
    });

    _last = EngineSnapshot(
      tsMs: now,
      bias: bias,
      longPct: longPct,
      shortPct: shortPct,
      consensus: consensus,
      confidence: conf,
      state: state,
      top: ev.take(6).toList(growable: false),
    );

    _out.add(_last);
  }
}
